### PR TITLE
[CN] Add implication keyword `implies`

### PR DESCRIPTION
**Pull request by elaustell**
Continuation of [#352](https://github.com/rems-project/cerberus/pull/352).

### CODE REVIEW

File: parsers/c/c_lexer.mll [hunk 1]
```diff
@@ -529,6 +529,7 @@ and initial flags = parse
   | "^="  { CARET_EQ            }
   | "|="  { PIPE_EQ             }
   | ','   { COMMA               }
+  | "==>" { EQ_EQ_GT            }
```
Comment (dc-mak) [inline parsers/c/c_lexer.mll]: Delete this

File: parsers/c/c_parser.mly [hunk 2]
```diff
@@ -122,7 +122,7 @@ type asm_qualifier =
 %token LBRACK RBRACK LPAREN RPAREN (*LBRACE RBRACE*) DOT MINUS_GT
   PLUS_PLUS MINUS_MINUS AMPERSAND STAR PLUS MINUS TILDE BANG
   SLASH PERCENT LT_LT GT_GT LT GT LT_EQ GT_EQ EQ_EQ BANG_EQ CARET PIPE
-  AMPERSAND_AMPERSAND PIPE_PIPE
+  AMPERSAND_AMPERSAND PIPE_PIPE EQ_EQ_GT
```
Comment (dc-mak) [inline parsers/c/c_parser.mly]: And this

File: parsers/c/c_parser.mly [hunk 3]
```diff
@@ -153,6 +153,7 @@ type asm_qualifier =
 %token CN_UNCHANGED CN_WILD CN_MATCH
 %token CN_GOOD CN_NULL CN_TRUE CN_FALSE
 %token <string * [`U|`I] * int> CN_CONSTANT
+%token CN_IMPLIES
```
Comment (dc-mak) [inline parsers/c/c_parser.mly]: Can you move this next to CN_TRUE and CN_FALSE?

File: parsers/c/tokens.ml [hunk 4]
```diff
@@ -112,6 +112,7 @@ type token =
   | PIPE_EQ
   | COMMA
   | LBRACK_LBRACK
+  | EQ_EQ_GT
```
Comment (dc-mak) [inline parsers/c/tokens.ml]: Please delete

File: parsers/c/tokens.ml [hunk 5]
```diff
@@ -281,6 +283,7 @@ let string_of_token = function
   | MINUS_EQ -> "MINUS_EQ"
   | LT_LT_EQ -> "LT_LT_EQ"
   | GT_GT_EQ -> "GT_GT_EQ"
+  | EQ_EQ_GT -> "EQ_EQ_GT"
```
Comment (dc-mak) [inline parsers/c/tokens.ml]: Please delete.

File: tests/cn/implies_associativity.c [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/* Here we test the right associativity of `implies`. Since `implies` is 
+   right associative, the expression `false implies true implies false` 
+   should be parsed as `false implies (true implies false)`,
+   and therefore should always be true. `(false implies true) implies false`
+   would fail.
+*/
+
+void foo ()
+{
+    /*@ assert(false implies true implies false); @*/
```
Comment (dc-mak) [inline tests/cn/implies_associativity.c:10]: Good test.

File: backend/cn/terms.ml [hunk 7]
```diff
@@ -189,8 +189,8 @@ let pp : 'bt 'a. ?atomic:bool -> ?f:('bt term -> Pp.document -> Pp.document) ->
           mparens (flow (break_op (ampersand ^^ ampersand)) [aux true it1; aux true it2])
        | Or ->
           mparens (flow (break_op (bar ^^ bar)) [aux true it1; aux true it2])
-       | Impl ->
-          mparens (flow (break_op (minus ^^ rangle ())) [aux true it1; aux true it2])
+       | Implies ->
+          mparens (flow (break_op (equals ^^ equals ^^ rangle ())) [aux true it1; aux true it2])
```
Comment (dc-mak) [inline backend/cn/terms.ml]: I think for now this should use the word "implies" just to stay consistent.

File: tests/cn/implies_precedence.c [hunk 8]
```diff
@@ -0,0 +1,10 @@
+/* Here we test the precedence of the implies operator. Since `and` is stronger than `implies`,
+   and `implies` is stronger than `or`, the expression `x == 0i32 && y == 0i32 || x==y implies y!=0i32 && x!=0i32`
+   should be parsed as `(x == 0i32 && y == 0i32) || (x==y implies (y!=0i32 && x!=0i32))`,
+   and therefore should always be true.
+*/
+
+int foo (int x, int y) {
+    /*@ assert (x == 0i32 && y == 0i32 || x==y implies y!=0i32 && x!=0i32); @*/
```
Comment (dc-mak) [inline tests/cn/implies_precedence.c:8]: Another good test, well done.

**Review by dc-mak — CHANGES_REQUESTED** [review]
Ok, final stretch. Thanks a lot for this work and your patience.
1. Just a few extraneous things leftover from previous versions - please delete. 
2. You'll need to change `backend/cn/cn_internal_to_ail.ml` and ask @rbanerjee20 for help to make sure the added tests don't break her stuff.
3. Once you're done these things, please rebase and squash all your changes once again, with (something like) the following commit message. Once that's done, push. I'll run the CI to make sure everything's ok, Rini will hopefully have confirmed by then that the changes are fine for her and we can rebase and commit onto Cerberus master :)

```
CN: Add `implies` logical binary operator for specs

This commit adds `implies` as a new, infix logical binary operator for
the spec language.  It does not implement it for the runtime testing.
The reason we use a new keyword "implies" instead of a more natural
`==>` is that currently the lexer is shared between C and CN, and and so
we cannot add `==>` to CN without adding it to C, which is incorrect
because C does not support that token.

Fixes https://github.com/rems-project/cerberus/issues/329
```

PS. See https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html and  https://chris.beams.io/posts/git-commit/#seven-rules for more details.

**Comment by rbanerjee20** [comment]
Looks good to me! Have taken a look at the change to `Cn_internal_to_ail.ml` and also have run the forked repo on the `austell_addImpl` branch locally - all tests for CN runtime testing are looking as they should.

**Pull request merged**